### PR TITLE
Add is_sparse function to array.hpp

### PR DESCRIPTION
**Pull request by herwinw**
This replaces the pattern `(AT(x) & SPARSE) != 0` often seen in the
code. This hides the actual implementation from the call sites.

### CODE REVIEW

File: jsrc/verbs/dyadic/take_drop.cpp [hunk 1]
```diff
@@ -201,8 +201,8 @@ jttake(J jt, A a, A w) {
     FPREFIP;
     if (!(a && w)) return 0;
     I wt = AT(w);  // wt=type of w
-    if ((SPARSE & AT(a)) != 0) RZ(a = jtdenseit(jt, a));
-    if (!(SPARSE & wt)) RZ(w = jtsetfv(jt, w, w));
+    if (is_sparse(a)) RZ(a = jtdenseit(jt, a));
+    if (!is_sparse(w)) RZ(w = jtsetfv(jt, w, w));
```
Comment (juntuu) [inline jsrc/verbs/dyadic/take_drop.cpp:205]: should check `wt`, not `w`

**Comment by herwinw** [comment]
Of course that one failing test passes locally, so I'm going to abuse the CI for a bit to find the source of the error.

**Comment by herwinw** [comment]
Well, it is solved by trying the same code again.So I guess something is not fully deterministic in the CI or the interpreter, but I don't think that's actually related to this code.

File: jsrc/array.hpp [hunk 2]
```diff
@@ -52,6 +52,11 @@ pointer_to_values(array x) -> int64_t* {
     return reinterpret_cast<int64_t*>(reinterpret_cast<C*>(x) + x->kchain.k);
 }
 
+[[nodiscard]] inline auto
+is_sparse(array x) -> bool {
```
Comment (codereport) [inline jsrc/array.hpp:56]: Should be `noexcept`

**Review by codereport — CHANGES_REQUESTED** [review]
will merge after this

**Comment by codereport** [comment]
also, need to resolve merge conflict. :pray:

File: jsrc/array.hpp (hunk 2 above)
Comment (codereport) [reply jsrc/array.hpp:56]: might as well add `constexpr`

Comment (herwinw) [reply jsrc/array.hpp]: Actually, has `constexpr` any effect here? The `array`/`A` argument is a pointer, that shouldn't be known at compile time.

Comment (codereport) [reply jsrc/array.hpp]: still good practice to put on all functions

**Review by codereport — APPROVED** [review]
:+1:

**Pull request merged**